Variant.String() panics if underlying type is []string

Using the convenience function `String()` when the underlying go type is not a literal `string` will result in a panic.

```go
func TestVariant(t *testing.T) {
	v := ua.MustVariant([]string{"a", "b", "c"})
	println(v.String())
}
```

```
--- FAIL: TestVariant (0.00s)
panic: interface conversion: interface {} is []string, not string [recovered]
	panic: interface conversion: interface {} is []string, not string
```

This is because we only check the opc type before casting and do not check array length or dimension.

At minimum we should change the convenience functions to return `($type, error)`. But I also think we should revisit the whole `Variant` interface since I am finding it a bit awkward to work with.

--- a/ua/variant.go
+++ b/ua/variant.go
@@ -458,6 +458,10 @@ func (m *Variant) set(v interface{}) error {
 // todo(fs): this should probably be StringValue or we need to handle all types
 // todo(fs): and recursion
 func (m *Variant) String() string {
+	if m.ArrayLength() > 0 {
+		return ""
+	}
+
 	switch m.Type() {
 	case TypeIDString:
 		return m.value.(string)
@@ -474,6 +478,10 @@ func (m *Variant) String() string {
 
 // Bool returns the boolean value if the type is Boolean.
 func (m *Variant) Bool() bool {
+	if m.ArrayLength() > 0 {
+		return false
+	}
+
 	switch m.Type() {
 	case TypeIDBoolean:
 		return m.value.(bool)
@@ -484,6 +492,10 @@ func (m *Variant) Bool() bool {
 
 // Float returns the float value if the type is one of the float types.
 func (m *Variant) Float() float64 {
+	if m.ArrayLength() > 0 {
+		return 0
+	}
+
 	switch m.Type() {
 	case TypeIDFloat:
 		return float64(m.value.(float32))
@@ -496,6 +508,10 @@ func (m *Variant) Float() float64 {
 
 // Int returns the int value if the type is one of the int types.
 func (m *Variant) Int() int64 {
+	if m.ArrayLength() > 0 {
+		return 0
+	}
+
 	switch m.Type() {
 	case TypeIDSByte:
 		return int64(m.value.(int8))
@@ -512,6 +528,10 @@ func (m *Variant) Int() int64 {
 
 // Uint returns the uint value if the type is one of the uint types.
 func (m *Variant) Uint() uint64 {
+	if m.ArrayLength() > 0 {
+		return 0
+	}
+
 	switch m.Type() {
 	case TypeIDByte:
 		return uint64(m.value.(byte))
@@ -527,6 +547,10 @@ func (m *Variant) Uint() uint64 {
 }
 
 func (m *Variant) ByteString() []byte {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDByteString:
 		return m.value.([]byte)
@@ -536,6 +560,10 @@ func (m *Variant) ByteString() []byte {
 }
 
 func (m *Variant) DataValue() *DataValue {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDDataValue:
 		return m.value.(*DataValue)
@@ -545,6 +573,10 @@ func (m *Variant) DataValue() *DataValue {
 }
 
 func (m *Variant) DiagnosticInfo() *DiagnosticInfo {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDDiagnosticInfo:
 		return m.value.(*DiagnosticInfo)
@@ -554,6 +586,10 @@ func (m *Variant) DiagnosticInfo() *DiagnosticInfo {
 }
 
 func (m *Variant) ExpandedNodeID() *ExpandedNodeID {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDExpandedNodeID:
 		return m.value.(*ExpandedNodeID)
@@ -563,6 +599,10 @@ func (m *Variant) ExpandedNodeID() *ExpandedNodeID {
 }
 
 func (m *Variant) ExtensionObject() *ExtensionObject {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDExtensionObject:
 		return m.value.(*ExtensionObject)
@@ -572,6 +612,10 @@ func (m *Variant) ExtensionObject() *ExtensionObject {
 }
 
 func (m *Variant) GUID() *GUID {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDGUID:
 		return m.value.(*GUID)
@@ -581,6 +625,10 @@ func (m *Variant) GUID() *GUID {
 }
 
 func (m *Variant) LocalizedText() *LocalizedText {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDLocalizedText:
 		return m.value.(*LocalizedText)
@@ -590,6 +638,10 @@ func (m *Variant) LocalizedText() *LocalizedText {
 }
 
 func (m *Variant) NodeID() *NodeID {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDNodeID:
 		return m.value.(*NodeID)
@@ -599,6 +651,10 @@ func (m *Variant) NodeID() *NodeID {
 }
 
 func (m *Variant) QualifiedName() *QualifiedName {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
+
 	switch m.Type() {
 	case TypeIDQualifiedName:
 		return m.value.(*QualifiedName)
@@ -608,6 +664,10 @@ func (m *Variant) QualifiedName() *QualifiedName {
 }
 
 func (m *Variant) StatusCode() StatusCode {
+	if m.ArrayLength() > 0 {
+		return StatusBadTypeMismatch
+	}
+
 	switch m.Type() {
 	case TypeIDStatusCode:
 		return m.value.(StatusCode)
@@ -618,6 +678,10 @@ func (m *Variant) StatusCode() StatusCode {
 
 // Time returns the time value if the type is DateTime.
 func (m *Variant) Time() time.Time {
+	if m.ArrayLength() > 0 {
+		return time.Time{}
+	}
+
 	switch m.Type() {
 	case TypeIDDateTime:
 		return m.value.(time.Time)
@@ -627,6 +691,9 @@ func (m *Variant) Time() time.Time {
 }
 
 func (m *Variant) Variant() *Variant {
+	if m.ArrayLength() > 0 {
+		return nil
+	}
 	switch m.Type() {
 	case TypeIDVariant:
 		return m.value.(*Variant)
@@ -636,6 +703,10 @@ func (m *Variant) Variant() *Variant {
 }
 
 func (m *Variant) XMLElement() XMLElement {
+	if m.ArrayLength() > 0 {
+		return ""
+	}
+
 	switch m.Type() {
 	case TypeIDXMLElement:
 		return m.value.(XMLElement)

--- a/ua/variant_test.go
+++ b/ua/variant_test.go
@@ -1046,9 +1046,135 @@ func TestVariantValueHelpers(t *testing.T) {
 			want: XMLElement("a"),
 			fn:   func(v *Variant) interface{} { return v.XMLElement() },
 		},
+
+		// []string
+		{
+			v:    []string{"a", "b", "c"},
+			want: "",
+			fn:   func(v *Variant) interface{} { return v.String() },
+		},
+
+		// []bool
+		{
+			v:    []bool{true, true, true},
+			want: false,
+			fn:   func(v *Variant) interface{} { return v.Bool() },
+		},
+
+		// []float64
+		{
+			v:    []float64{1, 2, 3},
+			want: float64(0),
+			fn:   func(v *Variant) interface{} { return v.Float() },
+		},
+
+		// []int64
+		{
+			v:    []int64{1, 2, 3},
+			want: int64(0),
+			fn:   func(v *Variant) interface{} { return v.Int() },
+		},
+
+		// []uint64
+		{
+			v:    []uint64{1, 2, 3},
+			want: uint64(0),
+			fn:   func(v *Variant) interface{} { return v.Uint() },
+		},
+
+		// [][]byte
+		{
+			v:    [][]byte{{'x', 'y', 'z'}},
+			want: ([]byte)(nil),
+			fn:   func(v *Variant) interface{} { return v.ByteString() },
+		},
+
+		// []*DataValue
+		{
+			v:    []*DataValue{{Status: StatusBad}},
+			want: (*DataValue)(nil),
+			fn:   func(v *Variant) interface{} { return v.DataValue() },
+		},
+
+		// []*DiagnosticInfo
+		{
+			v:    []*DiagnosticInfo{{AdditionalInfo: "nop"}},
+			want: (*DiagnosticInfo)(nil),
+			fn:   func(v *Variant) interface{} { return v.DiagnosticInfo() },
+		},
+
+		// []*ExpandedNodeID
+		{
+			v:    []*ExpandedNodeID{{NamespaceURI: "abc"}},
+			want: (*ExpandedNodeID)(nil),
+			fn:   func(v *Variant) interface{} { return v.ExpandedNodeID() },
+		},
+
+		// []*ExtensionObject
+		{
+			v:    []*ExtensionObject{{Value: "abc"}},
+			want: (*ExtensionObject)(nil),
+			fn:   func(v *Variant) interface{} { return v.ExtensionObject() },
+		},
+
+		// []*GUID
+		{
+			v:    []*GUID{NewGUID("abcd")},
+			want: (*GUID)(nil),
+			fn:   func(v *Variant) interface{} { return v.GUID() },
+		},
+
+		// []*LocalizedText
+		{
+			v:    []*LocalizedText{{Text: "abc"}},
+			want: (*LocalizedText)(nil),
+			fn:   func(v *Variant) interface{} { return v.LocalizedText() },
+		},
+
+		// []*NodeID
+		{
+			v:    []*NodeID{NewFourByteNodeID(1, 2)},
+			want: (*NodeID)(nil),
+			fn:   func(v *Variant) interface{} { return v.NodeID() },
+		},
+
+		// []*QualifiedName
+		{
+			v:    []*QualifiedName{{Name: "a"}},
+			want: (*QualifiedName)(nil),
+			fn:   func(v *Variant) interface{} { return v.QualifiedName() },
+		},
+
+		// []*StatusCode
+		{
+			v:    []StatusCode{StatusOK, StatusBad},
+			want: StatusBadTypeMismatch,
+			fn:   func(v *Variant) interface{} { return v.StatusCode() },
+		},
+
+		// []time.Time
+		{
+			v:    []time.Time{time.Date(2019, 1, 1, 12, 13, 14, 0, time.UTC)},
+			want: time.Time{},
+			fn:   func(v *Variant) interface{} { return v.Time() },
+		},
+
+		// []*Variant
+		{
+			v:    []*Variant{MustVariant("abc")},
+			want: (*Variant)(nil),
+			fn:   func(v *Variant) interface{} { return v.Variant() },
+		},
+
+		// []XMLElement
+		{
+			v:    []XMLElement{XMLElement("a")},
+			want: XMLElement(""),
+			fn:   func(v *Variant) interface{} { return v.XMLElement() },
+		},
 	}
-	for _, tt := range tests {
-		name := fmt.Sprintf("%T -> %T", tt.v, tt.want)
+	for i, tt := range tests {
+		name := fmt.Sprintf("test-%d %T -> %T", i, tt.v, tt.want)
 		t.Run(name, func(t *testing.T) {
 			verify.Values(t, "", tt.fn(MustVariant(tt.v)), tt.want)
 		})

